File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/provider/FileOperation.java
package app.zxtune.fs.provider;

import android.database.Cursor;
import android.database.MatrixCursor;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.provider.OpenableColumns;

import androidx.annotation.Nullable;

import java.io.File;
import java.io.IOException;

import app.zxtune.fs.Vfs;
import app.zxtune.fs.VfsFile;
import app.zxtune.fs.VfsObject;

class FileOperation implements AsyncQueryOperation {

  private static final String[] COLUMNS = {
      OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE};

  private final Uri uri;
  private final Resolver resolver;
  @Nullable
  private final String[] projection;

  FileOperation(Uri uri, Resolver resolver, @Nullable String[] projection) {
    this.uri = uri;
    this.resolver = resolver;
    this.projection = projection;
  }

  @Nullable
  @Override
  public Cursor call() throws Exception {
    final VfsFile file = maybeResolve();
    if (file == null) {
      return null;
    }
    final File f = maybeGetFile(file);
    if (f != null) {
      return makeResult(file, f);
    } else {
      return null;
    }
  }

  @Nullable
  private VfsFile maybeResolve() throws IOException {
    final VfsObject obj = resolver.resolve(uri);
    if (obj instanceof VfsFile) {
      return (VfsFile) obj;
    }
    return null;
  }

  @Nullable
  private File maybeGetFile(VfsFile file) {
    final File res = Vfs.getCacheOrFile(file);
    if (res != null && res.isFile()) {
      return res;
    }
    return null;
  }

  @Nullable
  @Override
  public Cursor status() {
    return null;
  }

  final ParcelFileDescriptor openFile(String mode) throws Exception {
    if (!"r".equals(mode)) {
      throw new IllegalArgumentException("Invalid mode: " + mode);
    }
    final VfsFile file = maybeResolve();
    if (file == null) {
      throw new IOException("Failed to resolve " + uri);
    }
    final File f = maybeGetFile(file);
    if (f == null) {
      throw new IOException("Failed to get file content of " + uri);
    }
    return ParcelFileDescriptor.open(f, ParcelFileDescriptor.MODE_READ_ONLY);
  }

  // as in FileProvider
  private Cursor makeResult(VfsFile file, File content) {
    final String[] columns = projection != null ? projection : COLUMNS;
    String[] cols = new String[columns.length];
    Object[] values = new Object[columns.length];
    int i = 0;
    for (String col : columns) {
      if (OpenableColumns.DISPLAY_NAME.equals(col)) {
        cols[i] = OpenableColumns.DISPLAY_NAME;
        values[i] = file.getName();
        ++i;
      } else if (OpenableColumns.SIZE.equals(col)) {
        cols[i] = OpenableColumns.SIZE;
        values[i] = content.length();
        ++i;
      }
    }

    cols = copyOf(cols, i);
    values = copyOf(values, i);

    final MatrixCursor cursor = new MatrixCursor(cols, 1);
    cursor.addRow(values);
    return cursor;
  }

  private static String[] copyOf(String[] original, int newLength) {
    final String[] result = new String[newLength];
    System.arraycopy(original, 0, result, 0, newLength);
    return result;
  }


  private static Object[] copyOf(Object[] original, int newLength) {
    final Object[] result = new Object[newLength];
    System.arraycopy(original, 0, result, 0, newLength);
    return result;
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/provider/FileOperation.kt
package app.zxtune.fs.provider

import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import android.provider.OpenableColumns
import androidx.annotation.VisibleForTesting
import app.zxtune.Logger
import app.zxtune.core.Identifier
import app.zxtune.core.jni.Api
import app.zxtune.fs.Vfs
import app.zxtune.fs.VfsFile
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.channels.WritableByteChannel

internal class FileOperation @VisibleForTesting constructor(
    uri: Uri,
    private val size: Long,
    private val resolver: Resolver,
    projection: Array<String>?,
    private val readData: (VfsFile) -> ByteBuffer,
    private val api: Api,
) : AsyncQueryOperation {

    private val id = Identifier(uri)
    private val columns = projection ?: COLUMNS

    constructor(uri: Uri, size: Long, resolver: Resolver, projection: Array<String>?) : this(
        uri, size, resolver, projection, Vfs::read, Api.instance()
    )

    override fun call() = maybeResolve()?.let {
        makeResult()
    }

    private fun maybeResolve() = resolver.resolve(id.dataLocation) as? VfsFile

    override fun status(): Cursor? = null

    fun consumeContent(out: WritableByteChannel) {
        val file = maybeResolve() ?: throw IOException("Failed to resolve $id")
        val rawData = readData(file)
        if (id.subPath.isEmpty()) {
            LOG.d { "Streaming ${size}/${rawData.limit()} bytes from $id" }
            out.write(rawData.asReadOnlyBuffer().limit(size.toInt()) as ByteBuffer)
        } else {
            api.loadModuleData(rawData, id.subPath) { moduleData ->
                LOG.d { "Streaming ${size}/${moduleData.capacity()} unpacked bytes from $id" }
                out.write(moduleData.asReadOnlyBuffer().limit(size.toInt()) as ByteBuffer)
            }
        }
    }

    private fun makeResult(): Cursor {
        val cols = arrayOfNulls<String>(columns.size)
        val values = arrayOfNulls<Any>(columns.size)
        var i = 0
        for (col in columns) {
            if (OpenableColumns.DISPLAY_NAME == col) {
                cols[i] = OpenableColumns.DISPLAY_NAME
                values[i] = id.virtualFilename
                ++i
            } else if (OpenableColumns.SIZE == col) {
                cols[i] = OpenableColumns.SIZE
                values[i] = size
                ++i
            }
        }
        return MatrixCursor(cols.copyOf(i), 1).apply {
            addRow(values.copyOf(i))
        }
    }

    companion object {
        private val COLUMNS = arrayOf(
            OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE
        )
        private val LOG = Logger(FileOperation::class.java.name)
    }
}
